File: prototype/rails/src/app/actions/carts/add_to_cart_test.rb
require "commerce_test_case"

module Carts
  class AddToCartTest < CommerceTestCase
    def test_it_puts_the_listing_in_the_cart
      art = listing(seller, quantity: 3)
      cart = cart_for(customer)

      item = AddToCart.new.call(cart: cart, listing: art, quantity: 2, now: moment("2026-08-20 08:00:00"))

      assert_equal 2, item.quantity
      assert_equal [art], cart.reload.items.map(&:listing)
    end

    def test_adding_the_same_listing_again_adds_to_the_line
      art = listing(seller, quantity: 3)
      cart = cart_for(customer)
      add_to_cart = AddToCart.new
      add_to_cart.call(cart: cart, listing: art, quantity: 1, now: moment("2026-08-20 08:00:00"))

      add_to_cart.call(cart: cart, listing: art, quantity: 1, now: moment("2026-08-20 08:05:00"))

      assert_equal 1, cart.reload.items.count
      assert_equal 2, cart.items.sole.quantity
    end

    def test_a_cart_never_holds_more_than_the_seller_has_left
      art = listing(seller, quantity: 2)

      item = AddToCart.new.call(cart: cart_for(customer), listing: art, quantity: 5, now: moment("2026-08-20 08:00:00"))

      assert_equal 2, item.quantity
    end

    def test_it_refuses_a_sold_out_listing
      art = listing(seller, quantity: 0, status: Domain::Listings::ListingStatus::SOLD)

      assert_raises(ArgumentError) do
        AddToCart.new.call(cart: cart_for(customer), listing: art, quantity: 1, now: moment("2026-08-20 08:00:00"))
      end
    end

    def test_it_records_the_interest_against_the_listing
      art = listing(seller)
      buyer = customer

      AddToCart.new.call(cart: cart_for(buyer), listing: art, quantity: 1, now: moment("2026-08-20 08:00:00"))

      event = art.listing_events.sole
      assert_equal Domain::Listings::ListingEventType::CART_ADD, event.event_type
      assert_equal buyer.id, event.customer_id
    end
  end
end

File: prototype/php/src/app/Actions/Cart/AddToCartTest.php
<?php

declare(strict_types=1);

namespace App\Actions\Cart;

use App\Domain\DomainRuleViolation;
use App\Domain\Listings\ListingEventType;
use App\Models\CustomerBlock;
use App\Models\ListingEvent;
use DomainException;

it('puts a listing in the cart', function (): void {
    $customer = $this->verifiedCustomer();
    $cart = $this->cartFor($customer);
    $listing = $this->listing($this->seller(), ['quantity' => 3]);

    $item = app(AddToCart::class)($cart, $listing, 2, $this->moment('2026-08-20 09:00:00'));

    expect($item->listing_id)->toBe($listing->id)
        ->and($item->quantity)->toBe(2);
});

it('raises the quantity on one line when the same listing is added twice', function (): void {
    $cart = $this->cartFor($this->verifiedCustomer());
    $listing = $this->listing($this->seller(), ['quantity' => 5]);
    $addToCart = app(AddToCart::class);
    $now = $this->moment('2026-08-20 09:00:00');

    $addToCart($cart, $listing, 1, $now);
    $item = $addToCart($cart, $listing, 2, $now);

    expect($item->quantity)->toBe(3)
        ->and($cart->items()->count())->toBe(1);
});

it('caps the quantity at the stock the listing has left', function (): void {
    $cart = $this->cartFor($this->verifiedCustomer());
    $listing = $this->listing($this->seller(), ['quantity' => 2]);

    $item = app(AddToCart::class)($cart, $listing, 9, $this->moment('2026-08-20 09:00:00'));

    expect($item->quantity)->toBe(2);
});

it('refuses a sold out listing', function (): void {
    $cart = $this->cartFor($this->verifiedCustomer());
    $listing = $this->listing($this->seller(), ['quantity' => 0]);

    app(AddToCart::class)($cart, $listing, 1, $this->moment('2026-08-20 09:00:00'));
})->throws(DomainException::class);

it('refuses a blocked customer', function (): void {
    $customer = $this->verifiedCustomer();
    CustomerBlock::factory()->create(['customer_id' => $customer->id, 'reason' => 'Chargeback fraud.']);
    $cart = $this->cartFor($customer);
    $listing = $this->listing($this->seller());

    $add = fn () => app(AddToCart::class)($cart, $listing, 1, $this->moment('2026-08-20 09:00:00'));

    expect($add)->toThrow(DomainRuleViolation::class, 'Chargeback fraud.')
        ->and($cart->items()->count())->toBe(0);
});

it('records the add as a listing event', function (): void {
    $customer = $this->verifiedCustomer();
    $cart = $this->cartFor($customer);
    $listing = $this->listing($this->seller());

    app(AddToCart::class)($cart, $listing, 1, $this->moment('2026-08-20 09:00:00'));

    $event = ListingEvent::query()->sole();

    expect($event->type)->toBe(ListingEventType::CartAdd)
        ->and($event->customer_id)->toBe($customer->id);
});
